File: Top-Interview-Questions-Leetcode/Easy-Collections/math/power-of-three.py
class Solution:
    def isPowerOfThree_recursion(self, n: int):
        if n <= 0:
            return False
        if n == 1:
            return True
        
        if n % 3 != 0:
            return False
        
        return self.isPowerOfThree_recursion(n // 3)
    
    def isPowerOfThree_binarySearch(self, n: int):
        if n <= 0:
            return False
        
        l, h = 0, 19
        
        while l <= h:
            m = l + (h - l) // 2
            val = pow(3, m) - n
            
            if val == 0:
                return True
            elif val > 0:
                h = m - 1
            else:
                l = m + 1
        
        return False
    
    def isPowerOfThree(self, n: int) -> bool:
        if n <= 0:
            return False
        
        return 1162261467 % n == 0

File: Top-Interview-Questions-Leetcode/Easy-Collections/math/power-of-three.cpp
#include <cmath>

class Solution {
public:
    
    bool isPowerOfThree_recursion(int n) {
        if (n <= 0) return false;
        if (n == 1) return true;
        if (n % 3 != 0) return false;
        
        return isPowerOfThree_recursion(n / 3);
    }
    
    bool isPowerOfThree_binarySearch(int n) {
        int l = 0, h = 19; // 3 ^ 20 > INT_MAX = 2^31 - 1
        
        while (l <= h) {
            int m = l + (h - l) / 2;
            int val = int(pow(3, m));
            
            if (n == val) return true;
            else if (n < val) h = m - 1;
            else l = m + 1;
        }
        
        return false;
    }
    
    bool isPowerOfThree(int n) {
        if (n <= 0) return false;
        
        return 1162261467 % n == 0;
    }
};